upgrade/upgrades: TestUpgradeSchemaChangerElements failed

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8744403?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8744403?buildTab=artifacts#/) on master @ [7e2df35a2f6bf7a859bb0539c8ca43c4e72ed260](https://github.com/cockroachdb/cockroach/commits/7e2df35a2f6bf7a859bb0539c8ca43c4e72ed260):


```
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2542 +0x50f
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130 +0x6ed
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1972 +0x7c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1977 +0x1d1a
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1895 +0x424
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1895 +0x424
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:864 +0x1be
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:731 +0x577

Goroutine 42416 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x619
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:332 +0x10d6
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob()
      github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:330 +0xc25
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs.func3()
      github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:210 +0xca
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs.func4()
      github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:213 +0x47

Goroutine 41786 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync()
      github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:644 +0x3c4
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:368 +0xd15
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:177 +0x48c
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:783 +0xc0a
  github.com/cockroachdb/cockroach/pkg/server.(*systemServerWrapper).serveConn()
      github.com/cockroachdb/cockroach/pkg/server/server_controller_sql.go:121 +0x1fa
  github.com/cockroachdb/cockroach/pkg/server.(*serverController).sqlMux()
      github.com/cockroachdb/cockroach/pkg/server/server_controller_sql.go:101 +0x51c
  github.com/cockroachdb/cockroach/pkg/server.(*serverController).sqlMux-fm()
      <autogenerated>:1 +0xc7
  github.com/cockroachdb/cockroach/pkg/server.startServeSQL.func1.1()
      github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1632 +0x2eb
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*TCPServer).ServeWith.func1()
      github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:188 +0x111
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x1f6
==================
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24608

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8779905?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8779905?buildTab=artifacts#/) on master @ [48065605ca2f44861b3d88c58b218acc28694ce4](https://github.com/cockroachdb/cockroach/commits/48065605ca2f44861b3d88c58b218acc28694ce4):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 335032 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestUpgradeSchemaChangerElements3762835689
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/schemachanger_elements_test.go
+++ b/pkg/upgrade/upgrades/schemachanger_elements_test.go
@@ -119,14 +119,14 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 				jobIsPaused                   chan struct{}
 				readyToQuery                  int64
 			)
-			scJobID := jobspb.InvalidJobID
+			scJobID := int64(jobspb.InvalidJobID)
 			params.Knobs.SQLDeclarativeSchemaChanger = &scexec.TestingKnobs{
 				BeforeStage: func(p scplan.Plan, stageIdx int) error {
 					if waitedForJob == nil {
 						return nil
 					}
 					if p.Stages[stageIdx].Phase == scop.PreCommitPhase {
-						scJobID = p.JobID
+						atomic.StoreInt64(&scJobID, int64(p.JobID))
 					} else if p.Stages[stageIdx].Phase > scop.PreCommitPhase {
 						<-waitedForJob
 						waitedForJob = nil
@@ -139,11 +139,11 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 			jobKnobs.IntervalOverrides.WaitForJobsInitialDelay = shortInterval()
 			jobKnobs.IntervalOverrides.WaitForJobsMaxDelay = shortInterval()
 			jobKnobs.BeforeWaitForJobsQuery = func(jobs []jobspb.JobID) {
-				if scJobID != jobspb.InvalidJobID {
+				if targetJobID := jobspb.JobID(atomic.LoadInt64(&scJobID)); targetJobID != jobspb.InvalidJobID {
 					for _, j := range jobs {
 						// The upgrade is waiting for the job...
-						if j == scJobID && atomic.LoadInt64(&readyToQuery) > 0 {
-							scJobID = jobspb.InvalidJobID
+						if j == targetJobID && atomic.LoadInt64(&readyToQuery) > 0 {
+							atomic.StoreInt64(&scJobID, int64(jobspb.InvalidJobID))
 							atomic.StoreInt64(&readyToQuery, 0)
 							waitedForJob <- struct{}{}
 						}
@@ -180,7 +180,7 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 				if tc.unpauseJob {
 					_, err = sqlDB.Exec("SET CLUSTER SETTING jobs.debug.pausepoints=''")
 					require.NoError(t, err)
-					_, err = sqlDB.Exec("RESUME JOB $1", scJobID)
+					_, err = sqlDB.Exec("RESUME JOB $1", jobspb.JobID(atomic.LoadInt64(&scJobID)))
 					require.NoError(t, err)
 				}
 				_, err = sqlDB.Exec("SET CLUSTER SETTING version = crdb_internal.node_executable_version()")

